File: python/Decorator/StarbuzzCoffee/beverage.py
from abc import ABC, abstractmethod


class Beverage(ABC):
    @abstractmethod
    def getDescription(self) -> str:
        pass

    @abstractmethod
    def cost(self) -> float:
        pass


class Expresso(Beverage):
    def __init__(self):
        self.description = "Expresso"
        self.price = 1.99

    def getDescription(self) -> str:
        return self.description

    def cost(self) -> float:
        return self.price


class HouseBlend(Beverage):
    def __init__(self):
        self.description = "House Blend Coffee"
        self.price = 0.89

    def getDescription(self) -> str:
        return self.description

    def cost(self) -> float:
        return self.price


class DarkRoast(Beverage):
    def __init__(self):
        self.description = "Dark Roast Coffee"
        self.price = 0.99

    def getDescription(self) -> str:
        return self.description

    def cost(self) -> float:
        return self.price

File: c/Decorator/Beverage.c
#include "Beverage.h"

void ExpressoNew(Expresso *this)
{
    strcpy(this->description, "Expresso");
    this->price = 1.99;
    this->getDescription = ExpGetDescription;
    this->cost = ExpCost;
}
char *ExpGetDescription(Expresso *this)
{
    return this->description;
}
float ExpCost(Expresso *this)
{
    return this->price;
}

void HouseBlendNew(HouseBlend *this)
{
    strcpy(this->description, "HouseBlend");
    this->price = 0.89;
    this->getDescription = HBGetDescription;
    this->cost = HBCost;
}
char *HBGetDescription(HouseBlend *this)
{
    return this->description;
}
float HBCost(HouseBlend *this)
{
    return this->price;
}

void DarkRoastNew(DarkRoast *this)
{
    strcpy(this->description, "DarkRoast");
    this->price = 0.99;
    this->getDescription = DRGetDescription;
    this->cost = DRCost;
}
char *DRGetDescription(DarkRoast *this)
{
    return this->description;
}
float DRCost(DarkRoast *this)
{
    return this->price;
}